File: server/auth/local/local.py
import secrets
import sys
from base64 import b32encode
from datetime import datetime, timedelta, timezone

from fastapi import Depends, HTTPException, Request
from fastapi.security import OAuth2PasswordBearer
from jose import JWTError, jwt
from pyotp import TOTP
from pyotp.utils import build_uri
from qrcode import QRCode

from global_config import AuthType
from helpers import get_env, verify_password
from logger import logger

from ..base import BaseAuth
from ..models import Login, Token

oauth2_scheme = OAuth2PasswordBearer(tokenUrl="api/token", auto_error=False)


class LocalAuth(BaseAuth):
    JWT_ALGORITHM = "HS256"

    def __init__(self, global_config) -> None:
        # Credentials come from environment variables when set (env always
        # wins), otherwise from the stored first-run setup config.
        stored_config = global_config.stored_config or {}
        self.username = get_env("GLOBNOTES_USERNAME") or stored_config.get(
            "username"
        )
        self.password = get_env("GLOBNOTES_PASSWORD")
        self.password_hash = (
            None if self.password else stored_config.get("password_hash")
        )
        self.secret_key = get_env("GLOBNOTES_SECRET_KEY") or stored_config.get(
            "secret_key"
        )
        if (
            not self.username
            or not (self.password or self.password_hash)
            or not self.secret_key
        ):
            logger.error(
                "Login credentials must be provided via environment "
                "variables or the first-run setup wizard."
            )
            sys.exit(1)
        self.username = self.username.lower()
        self.session_expiry_days = get_env(
            "GLOBNOTES_SESSION_EXPIRY_DAYS", default=30, cast_int=True
        )

        # TOTP (env-configured only)
        self.is_totp_enabled = False
        if global_config.auth_type == AuthType.TOTP:
            if not self.password:
                logger.error(
                    "GLOBNOTES_PASSWORD must be set when using TOTP auth."
                )
                sys.exit(1)
            self.is_totp_enabled = True
            self.totp_key = get_env("GLOBNOTES_TOTP_KEY", mandatory=True)
            self.totp_key = b32encode(self.totp_key.encode("utf-8"))
            self.totp = TOTP(self.totp_key)
            self.last_used_totp = None
            self._display_totp_enrolment()

    def login(self, data: Login) -> Token:
        # Check Username
        username_correct = secrets.compare_digest(
            self.username.lower(), data.username.lower()
        )

        # Check Password & TOTP
        if self.is_totp_enabled:
            current_totp = self.totp.now()
            password_correct = secrets.compare_digest(
                self.password + current_totp, data.password
            )
        elif self.password is not None:
            password_correct = secrets.compare_digest(
                self.password, data.password
            )
        else:
            password_correct = verify_password(
                data.password, self.password_hash
            )

        # Raise error if incorrect
        if not (
            username_correct
            and password_correct
            # Prevent TOTP from being reused
            and (
                self.is_totp_enabled is False
                or current_totp != self.last_used_totp
            )
        ):
            raise ValueError("Incorrect login credentials.")
        if self.is_totp_enabled:
            self.last_used_totp = current_totp

        # Create Token
        access_token = self._create_access_token(data={"sub": self.username})
        return Token(access_token=access_token)

    def authenticate(
        self, request: Request, token: str = Depends(oauth2_scheme)
    ):
        # If no token is found in the header, check the cookies
        if token is None:
            token = request.cookies.get("token")
        # Validate the token
        try:
            self._validate_token(token)
        except (JWTError, ValueError):
            raise HTTPException(
                status_code=401,
                detail="Invalid authentication credentials",
                headers={"WWW-Authenticate": "Bearer"},
            )

    def _validate_token(self, token: str) -> bool:
        if token is None:
            raise ValueError
        payload = jwt.decode(
            token, self.secret_key, algorithms=[self.JWT_ALGORITHM]
        )
        username = payload.get("sub")
        if username is None or username.lower() != self.username:
            raise ValueError

    def _create_access_token(self, data: dict):
        to_encode = data.copy()
        expiry_datetime = datetime.now(timezone.utc) + timedelta(
            days=self.session_expiry_days
        )
        to_encode.update({"exp": expiry_datetime})
        encoded_jwt = jwt.encode(
            to_encode, self.secret_key, algorithm=self.JWT_ALGORITHM
        )
        return encoded_jwt

    def _display_totp_enrolment(self):
        # Fix for #237. Remove padding as per spec:
        # https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret
        unpadded_secret = self.totp_key.rstrip(b"=")
        uri = build_uri(unpadded_secret, self.username, issuer="globnotes")
        qr = QRCode()
        qr.add_data(uri)
        print(
            "\nScan this QR code with your TOTP app of choice",
            "e.g. Authy or Google Authenticator:",
        )
        qr.print_ascii()
        print(
            f"Or manually enter this key: {self.totp.secret.decode('utf-8')}\n"
        )

File: server/auth/local.ts
// SPDX-License-Identifier: LGPL-3.0-only

/**
 * Local username/password (and optional TOTP) authentication.
 * Ported from the Python server's auth/local/local.py with identical
 * semantics: env credentials win over stored config, HS256 JWTs, TOTP
 * codes append to the plaintext password, TOTP codes are single-use.
 */

import { jwtVerify, SignJWT } from "jose";
import { totp } from "otplib";
import { encodeBase32 } from "@std/encoding/base32";
import QRCode from "qrcode";

import { AuthType, type GlobalConfig } from "../config.ts";
import { getEnv, timingSafeEqual, verifyPassword } from "../helpers.ts";
import { logger } from "../logger.ts";
import type { Login, Token } from "./models.ts";

const JWT_ALGORITHM = "HS256";

export class LocalAuth {
  readonly username: string;
  /** Plaintext password from env (takes precedence over the stored hash). */
  private readonly password: string | null;
  private readonly passwordHash: string | null;
  private readonly secretKey: string;
  private readonly sessionExpiryDays: number;
  readonly isTotpEnabled: boolean;
  private totpSecret = "";
  private lastUsedTotp: string | null = null;

  constructor(globalConfig: GlobalConfig) {
    // Credentials come from environment variables when set (env always
    // wins), otherwise from the stored first-run setup config.
    const stored = globalConfig.storedConfig ?? {};
    this.username = (getEnv("GLOBNOTES_USERNAME") || stored.username || "")
      .toLowerCase();
    this.password = getEnv("GLOBNOTES_PASSWORD") || null;
    this.passwordHash = this.password ? null : (stored.password_hash ?? null);
    this.secretKey = getEnv("GLOBNOTES_SECRET_KEY") || stored.secret_key || "";
    if (
      !this.username || !(this.password || this.passwordHash) || !this.secretKey
    ) {
      logger.error(
        "Login credentials must be provided via environment " +
          "variables or the first-run setup wizard.",
      );
      Deno.exit(1);
    }
    this.sessionExpiryDays = Number(
      getEnv("GLOBNOTES_SESSION_EXPIRY_DAYS", { castInt: true, default: 30 }),
    );

    // TOTP (env-configured only)
    this.isTotpEnabled = false;
    if (globalConfig.authType === AuthType.TOTP) {
      if (!this.password) {
        logger.error("GLOBNOTES_PASSWORD must be set when using TOTP auth.");
        Deno.exit(1);
      }
      this.isTotpEnabled = true;
      // The env value is raw text; the server base32-encodes it (same as
      // pyotp's b32encode in the Python server).
      this.totpSecret = encodeBase32(
        new TextEncoder().encode(
          getEnv("GLOBNOTES_TOTP_KEY", { mandatory: true }),
        ),
      );
    }
  }

  async login(data: Login): Promise<Token> {
    const usernameCorrect = timingSafeEqual(
      this.username,
      (data.username ?? "").toLowerCase(),
    );

    let passwordCorrect: boolean;
    let currentTotp: string | null = null;
    if (this.isTotpEnabled) {
      currentTotp = totp.generate(this.totpSecret);
      passwordCorrect = timingSafeEqual(
        this.password + currentTotp,
        data.password ?? "",
      );
    } else if (this.password !== null) {
      passwordCorrect = timingSafeEqual(this.password, data.password ?? "");
    } else {
      passwordCorrect = await verifyPassword(
        data.password ?? "",
        this.passwordHash!,
      );
    }

    if (
      !(usernameCorrect && passwordCorrect &&
        (!this.isTotpEnabled || currentTotp !== this.lastUsedTotp))
    ) {
      throw new Error("Incorrect login credentials.");
    }
    if (this.isTotpEnabled) {
      this.lastUsedTotp = currentTotp;
    }

    return {
      access_token: await this.#createAccessToken(),
      token_type: "bearer",
    };
  }

  /** Validate a bearer token; throws on missing/invalid/expired. */
  async validateToken(token: string | null | undefined): Promise<void> {
    if (!token) throw new Error("no token");
    const key = new TextEncoder().encode(this.secretKey);
    const { payload } = await jwtVerify(token, key);
    const sub = payload.sub;
    if (!sub || sub.toLowerCase() !== this.username) {
      throw new Error("wrong subject");
    }
  }

  async #createAccessToken(): Promise<string> {
    const key = new TextEncoder().encode(this.secretKey);
    const exp = Math.floor(Date.now() / 1000) +
      this.sessionExpiryDays * 86_400;
    return await new SignJWT({ sub: this.username })
      .setProtectedHeader({ alg: JWT_ALGORITHM })
      .setExpirationTime(exp)
      .sign(key);
  }

  /** Print the TOTP enrolment QR code + manual key at startup
   * (same as the Python server). Call after construction. */
  async displayTotpEnrolment(): Promise<void> {
    const unpaddedSecret = this.totpSecret.replace(/=+$/, "");
    const uri =
      `otpauth://totp/globnotes:${encodeURIComponent(this.username)}` +
      `?secret=${unpaddedSecret}&issuer=globnotes`;
    const qr = await QRCode.toString(uri, { type: "terminal", small: true });
    console.log(
      "\nScan this QR code with your TOTP app of choice",
      "e.g. Authy or Google Authenticator:\n",
      qr,
      `\nOr manually enter this key: ${unpaddedSecret}\n`,
    );
  }
}
